enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go

```diff
@@ -395,26 +416,49 @@ func (q *PriorityTaskScheduler) stats() string {
 		len(q.activeTaskCancelFuncs), q.q.Len())
 }
 
-func (q *PriorityTaskScheduler) canFitAnotherTask(res *scpb.EnqueueTaskReservationRequest) bool {
-	// Only ever run as many sized tasks as we have memory for.
-	knownRAMremaining := q.ramBytesCapacity - q.ramBytesUsed
-	knownCPUremaining := q.cpuMillisCapacity - q.cpuMillisUsed
-	willFit := knownRAMremaining >= res.GetTaskSize().GetEstimatedMemoryBytes() && knownCPUremaining >= res.GetTaskSize().GetEstimatedMilliCpu()
+func (q *PriorityTaskScheduler) canFitTask(res *scpb.EnqueueTaskReservationRequest) bool {
+	// If we're running in exclusiveTaskScheduling mode, only ever allow one
+	// task to run at a time. Otherwise fall through to the logic below.
+	if q.exclusiveTaskScheduling && len(q.activeTaskCancelFuncs) >= 1 {
+		return false
+	}
+
+	size := res.GetTaskSize()
+
+	availableRAM := q.ramBytesCapacity - q.ramBytesUsed
+	if size.GetEstimatedMemoryBytes() > availableRAM {
+		return false
+	}
 
-	// If we're running in exclusiveTaskScheduling mode, only ever allow one task to run at
-	// a time. Otherwise fall through to the logic below.
-	if willFit && q.exclusiveTaskScheduling && len(q.activeTaskCancelFuncs) >= 1 {
+	availableCPU := q.cpuMillisCapacity - q.cpuMillisUsed
+	if size.GetEstimatedMilliCpu() > availableCPU {
 		return false
 	}
 
-	if willFit {
-		if res.GetTaskSize().GetEstimatedMemoryBytes() == 0 {
-			log.CtxWarningf(q.rootContext, "Scheduling another unknown size task. THIS SHOULD NOT HAPPEN! res: %+v", res)
-		} else {
-			log.CtxInfof(q.rootContext, "Scheduling another task of size: %+v", res.GetTaskSize())
+	for _, r := range size.GetCustomResources() {
+		used, ok := q.customResourcesUsed[r.GetName()]
+		if !ok {
+			// The scheduler should never send us tasks that require resources
+			// we haven't set up in the config.
+			alert.UnexpectedEvent("missing_custom_resource", "Task requested custom resource %q which is not configured for this executor", r.GetName())
+			continue
+		}
+		available := q.customResourcesCapacity[r.GetName()] - used
+		if customResource(r.GetValue()) > available {
+			return false
 		}
 	}
-	return willFit
+
+	// The scheduler should prevent CPU/memory requests that are <= 0.
+	// Alert if we get a task like this.
+	if size.GetEstimatedMemoryBytes() <= 0 {
+		alert.UnexpectedEvent("invalid_task_memory", "Requested memory %d is invalid", size.GetEstimatedMemoryBytes())
```

Comment:
ditto here, isn't this just a client-specified string? (and below)

Reply:
If the client asks for CPU or memory < 0, then it shouldn't make it this far due to the logic [here](https://github.com/buildbuddy-io/buildbuddy/blob/43174a2ff85e3047ca5860f040d5e004b2e3d850/enterprise/server/tasksize/tasksize.go#L399)